src/main/java/com/fasterxml/jackson/databind/type/TypeBindings.java

```diff
@@ -253,6 +250,24 @@ public JavaType findBoundType(String name)
         return null;
     }
 
+    /**
+     * Returns true if a shallow search of the type bindings includes a placeholder
+     * type which uses reference equality, thus cannot produce cache hits. This
+     * is an optimization to avoid churning memory in the cache unnecessarily.
+     * Note that it is still possible for nested type information to contain such
+     * placeholder types (see NestedTypes1604Test for an example) so it's vital
+     * that they produce a distribution of hashCode values, even if they may push
+     * reusable data out of the cache.
+     */
+    private boolean invalidCacheKey() {
```

Comment:
@cowtowncoder I've updated this PR with commit eb5cc8bf29161c9076e47334681a9ac0cd601334, the above comment describes the rationale for making `JavaType.hashCode` non-final, due to the example in NestedTypes1604Test where we don't catch all cases. I've updated this shallow check to match IdentityEqualityType which covers both `PlaceholderForType` and `ResolvedRecursiveType`, both of which use reference equality.